staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go

```diff
@@ -189,17 +193,17 @@ type udsGRPCConnector struct {
 	udsName string
 }
 
-func (u *udsGRPCConnector) connect() (proxier, error) {
+func (u *udsGRPCConnector) connect(ctx context.Context) (proxier, error) {
 	udsName := u.udsName
-	dialOption := grpc.WithContextDialer(func(context.Context, string) (net.Conn, error) {
-		c, err := net.Dial("unix", udsName)
+	dialOption := grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+		var d net.Dialer
+		c, err := d.DialContext(ctx, "unix", udsName)
 		if err != nil {
 			klog.Errorf("failed to create connection to uds name %s, error: %v", udsName, err)
 		}
 		return c, err
 	})
 
-	ctx := context.TODO()
```

Comment:
cc @cheftako @jkh52 

I'm unsure about this. I asked for the same thing in https://github.com/kubernetes/kubernetes/pull/103126/files#r658721685, but there were some weird implications with a request-level context getting tied to a proxy connection that ended up getting used for multiple requests

Reply:
Context is only used for the dial, connection lifetime is not tied to it:

- https://github.com/golang/go/blob/1170771074c0792f36fdfe94e50de1dcc8144946/src/net/dial.go#L361-L363
- https://github.com/golang/go/blob/1170771074c0792f36fdfe94e50de1dcc8144946/src/crypto/tls/tls.go#L211-L213

I don't see any reason for concern, am I missing something?

Reply:
Sorry trying to page that in to my memory.  What we discovered with https://github.com/kubernetes/kubernetes/pull/103272 (an earlier version of this fix)  was problems with pull-kubernetes-e2e-gce-ubuntu-containerd. This version seems to have the same issues. My memory is that the context was being linked with the tunnel. For certain long lived operations having to do with with storage we were seeing requests failing because the context had timed out, there by killing the underlying tunnel. However my memory from a year ago maybe slightly off.

Reply:
Ah, I looked at [the code](https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/eb4ec1a7ef321b28eb635ba8d9e704b4dfb2c183/konnectivity-client/pkg/client/client.go#L76-L100) and I see what the problems is. Both the dial and the RPC use the same context, so using the context from a concrete request is indeed incorrect. It could be used for dial only, but there is no way to do that. There is also no way to control the lifecycle of the proxy connection from the outside, at least I don't see how the called would shut the tunnel down. This needs to be addressed in the konnectivity-client itself. I put the TODO context back and left a comment there.

Reply:
The right approach seem to be to call `grpc.DialContext()` within `CreateSingleUseGrpcTunnel()` but `grpcClient.Proxy(ctx)` should be moved into the `grpcTunnel.DialContext()`. That way the `Tunnel` that `CreateSingleUseGrpcTunnel()` returns becomes reusable rather than single use and it's whole lifetime can be controlled via the context passed to `CreateSingleUseGrpcTunnel()`.

Happy to open a PR if the above makes sense.

Reply:
> This needs to be addressed in the konnectivity-client itself. I put the TODO context back and left a comment there.

can you copy the relevant details into an issue in that repo and link to it from the TODO in the code here?

Reply:
Opened an issue here https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/357 and updated the comment.